- Description: zfcp: downloading generic ACT leads to kernel BUG
  Symptom:     kernel BUG: illegal operation
  Problem:     condition in BUG_ON statement was too strict
  Solution:    changed condition in BUG_ON statement
  Problem-ID:  
  Archs:       s390-31, s390-64

--- linux-2.5/drivers/s390/scsi/zfcp_aux.c	19 May 2004 07:40:20 -0000	1.105.2.2
+++ linux-2.5/drivers/s390/scsi/zfcp_aux.c	16 Jun 2004 16:03:07 -0000	1.105.2.3
@@ -29,7 +29,7 @@
  */
 
 /* this drivers version (do not edit !!! generated and updated by cvs) */
-#define ZFCP_AUX_REVISION "$Revision: 1.105.2.2 $"
+#define ZFCP_AUX_REVISION "$Revision: 1.105.2.3 $"
 
 #include "zfcp_ext.h"
 
@@ -427,7 +427,7 @@
 		retval = -ENOMEM;
 		goto out;
 	}
-	sg_list->count = 0;
+	memset(sg_list, 0, sizeof(*sg_list));
 
 	if (command != ZFCP_CFDC_IOC) {
 		ZFCP_LOG_INFO("IOC request code 0x%x invalid\n", command);
@@ -612,6 +612,7 @@
 	sg_list->sg = kmalloc(sg_list->count * sizeof(struct scatterlist),
 			      GFP_KERNEL);
 	if (sg_list->sg == NULL) {
+		sg_list->count = 0;
 		retval = -ENOMEM;
 		goto out;
 	}
@@ -648,10 +649,12 @@
 	unsigned int i;
 	int retval = 0;
 
-	BUG_ON((sg_list->sg == NULL) || (sg_list == NULL));
+	BUG_ON(sg_list == NULL);
 
 	for (i = 0, sg = sg_list->sg; i < sg_list->count; i++, sg++)
 		__free_pages(sg->page, 0);
+
+	kfree(sg_list->sg);
 
 	return retval;
 }

